module/zfs/dmu_recv.c

```diff
@@ -2274,7 +2645,8 @@ dmu_recv_cleanup_ds(dmu_recv_cookie_t *drc)
 		rrw_exit(&ds->ds_bp_rwlock, FTAG);
 		dsl_dataset_name(ds, name);
 		dsl_dataset_disown(ds, dsflags, dmu_recv_tag);
-		(void) dsl_destroy_head(name);
+		if (!drc->drc_heal)
```

Comment:
we skip the destroy because we operated on an existing snapshot instead of the temp clone. I've updated the comment here: https://github.com/openzfs/zfs/pull/9372/files#diff-aa0a401dc1077010f8856e9acf77f4dcR848 to make this clearer.